src/main/java/com/puppycrawl/tools/checkstyle/utils/XpathUtil.java

```diff
@@ -0,0 +1,117 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.utils;
+
+import java.util.Arrays;
+import java.util.List;
+
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * Contains utility methods for xpath.
+ *
+ */
+public final class XpathUtil {
+
+    /**
+     * List of token types which support value attribute.
+     * These token types were selected based on analysis that all others do not match required
+     * criteria - value of the token must be useful and help to retrieve more precise results.
+     * There are three types of AST tokens:
+     * 1. Tokens for which the values are equal to the name of the token. Or in other words,
+     * nodes for which the following expression is always true:
+     * <pre>
+     *     detailAst.getText().equals(TokenUtil.getTokenName(detailAst.getType()))
+     * </pre>
+     * For example:
+     * <pre>
+     *     //MODIFIERS[@text='MODIFIERS']
+     *     //OBJBLOCK[@text='OBJBLOCK']
+     * </pre>
+     * These tokens do not match required criteria because their values do not carry any
+     * information, they do not affect the xpath requests and do not help to get more accurate
+     * results. The values of these nodes are useless. No matter which class you analyze, these
+     * values are always the same.
+     * In addition, they make xpath queries more complex, less readable and less obvious.
+     * 2. Tokens for which the values differ from token names, but values are always constant.
+     * For example:
+     * <pre>
+     *     //LITERAL_VOID[@text='void']
+     *     //RCURLY[@text='}']
+     * </pre>
+     * These tokens are not used for the same reasons as were described in the previous part.
+     * 3. Tokens for which values are not constant. The values of these nodes are closely related
+     * to a concrete class, method, variable and so on.
+     * For example:
+     * <pre>
+     *     String greeting = "HelloWorld";
+     *     //STRING_LITERAL[@text='HelloWorld']
+     * </pre>
+     * <pre>
+     *     int year = 2017;
+     *     //NUM_INT[@text=2017]
+     * </pre>
+     * <pre>
+     *     int age = 23;
+     *     //NUM_INT[@text=23]
+     * </pre>
+     * As you can see same {@code NUM_INT} token type can have different values, depending on
+     * context.
+     * <pre>
+     *     public class MyClass {}
+     *     //IDENT[@text='MyClass']
+     * </pre>
+     * Only these tokens support value attribute because they make our xpath queries more accurate.
```

Comment:
`value attribute` ?? do you mean "text atttribute"
please review my comment below.

Reply:
done